src/methods.jl

```diff
@@ -10,318 +19,451 @@ derivatives.
-    p = length(grid)
-    _check_p_q(p, q)
-    coefs = _coefs(grid, p, q)
-    hist = History(; adapt=adapt, kwargs...)
-    return Nonstandard{typeof(grid), typeof(coefs)}(Int(p), Int(q), grid, coefs, hist)
+function (m::FiniteDifferenceMethod)(f::Function, x::T, h) where T<:AbstractFloat
+    return sum(
+        i -> convert(T, m.coefs[i]) * f(T(x + h * m.grid[i])),
+        eachindex(m.grid)
+    ) / h^m.q
 end
+# Handle arguments that are not floats. Assume that converting to float is desired.
+(m::FiniteDifferenceMethod)(f::Function, x::T, h) where T<:Real = m(f, float(x), h)
 
-# Check the method and derivative orders for consistency
+# Check the method and derivative orders for consistency.
 function _check_p_q(p::Integer, q::Integer)
-    q >= 0 || throw(ArgumentError("order of derivative must be nonnegative"))
-    q < p || throw(ArgumentError("order of the method must be strictly greater than that " *
-                                 "of the derivative"))
-    # Check whether the method can be computed. We require the factorial of the
-    # method order to be computable with regular `Int`s, but `factorial` will overflow
-    # after 20, so 20 is the largest we can allow.
-    p > 20 && throw(ArgumentError("order of the method is too large to be computed"))
+    q >= 0 || throw(DomainError(q, "order of derivative (`q`) must be non-negative"))
+    q < p || throw(DomainError(
+        (q, p),
+        "order of the method (q) must be strictly greater than that of the derivative (p)",
+    ))
+    # Check whether the method can be computed. We require the factorial of the method order
+    # to be computable with regular `Int`s, but `factorial` will after 20, so 20 is the
+    # largest we can allow.
+    p > 20 && throw(DomainError(p, "order of the method (`p`) is too large to be computed"))
     return
 end
 
-# Compute coefficients for the method
+const _COEFFS_CACHE = Dict{Tuple{AbstractVector{<:Real}, Integer}, Vector{Float64}}()
 
-const _COEFFS_CACHE = Dict{Tuple{AbstractVector{<:Real}, Integer, Integer}, Vector{Float64}}()
-function _coefs(grid::AbstractVector{<:Real}, p::Integer, q::Integer)
-    return get!(_COEFFS_CACHE, (grid, p, q)) do
-        # For high precision on the \ we use Rational, and to prevent overflows we use Int128
-        # At the end we go to Float64 for fast floating point math (rather than rational math)
+# Compute coefficients for the method and cache the result.
+function _coefs(grid::AbstractVector{<:Real}, q::Integer)
+    return get!(_COEFFS_CACHE, (grid, q)) do
+        p = length(grid)
+        # For high precision on the `\`, we use `Rational`, and to prevent overflows we use
+        # `Int128`. At the end we go to `Float64` for fast floating point math, rather than
+        # rational math.
         C = [Rational{Int128}(g^i) for i in 0:(p - 1), g in grid]
         x = zeros(Rational{Int128}, p)
         x[q + 1] = factorial(q)
         return Float64.(C \ x)
```

Comment:
Should it always be `Float64` or sould it be based on `eltype(grid)`?

Reply:
The grid can be integers, so `eltype(grid)` won't always work. It should be the floating point type for the step size computation. `Float64` seems like a reasonable compromise. (I actually pretty much copied this piece of the code from the current `methods.jl`.) We could even make this part of the `FiniteDifferenceMethod` type, to e.g. allow step size computations with `Float64`s or even `Rational`s even though the input to the function is e.g. `Float32`.

Reply:
You could use `float(eltype(grid))`... if the user is employing more than double precision, it seems a shame to discard it.

Reply:
I guess lets leave it as Float64 for now.
(I wrote it so its not like it has to be right)

Reply:
I like `float(eltype(grid))`, as it would indeed allow the user to specify a higher precision. It does feel a bit strange, however, that the user then is required to specify the data type of internal computations by the data type of the grid. It seems more logical to have it be a type parameter of `FiniteDifferenceMethod`.

Reply:
hmm because we cache it by the value of the grid this could lead to weirdness.
since `isequal(Float32.(grid), Float64.(grid))`

Lets leave it for a future PR?

Reply:
Yes, let's indeed leave it for a future PR.